src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -249,6 +388,107 @@ struct MHAKernel<ScaledDotProductAttention::KT_ONEDNN, T> {
+                                     &qk_scratch_b.at<bfloat16>({b, h / h_each_group_len, 0}));
+            float* alibi_ptr = nullptr;
+            auto alibi_stride = 0;
+            if (alibi_mask) {
+                alibi_ptr = &alibi_mask.at<float>({b, h, 0, 0}, true);
+                if (alibi_mask.size(2) > 1)
+                    alibi_stride = alibi_mask.stride(2);
+            }
+            auto attn_mask_prec = attention_mask.get_precision();
+            uint8_t* attn_mask_ptr = nullptr;
+            auto attn_mask_stride = 0;
+            if (attention_mask) {
+                attn_mask_ptr = get_attn_mask_ptr(attention_mask, b, h);
+                if (attention_mask.size(2) > 1)
+                    attn_mask_stride = attention_mask.stride(2) * attn_mask_prec.size();
+            }
+            uint8_t* cmask_ptr = nullptr;
+            auto cmask_stride = 0;
+            if (causal_mask) {
+                cmask_ptr = &causal_mask.at<uint8_t>({b, h, 0, 0}, true);
+                if (causal_mask.size(2) > 1)
+                    cmask_stride = causal_mask.stride(2);
+            }
+            for (size_t m = m_start; m < m_end; m++) {
+                // apply attention mask & sofmax
+                auto ncausal = auto_causal ? (kv_len - q_len + m + 1) : kv_len;
+                attn_softmax(&score.at<float>({b, h, m, 0}),
+                             &weight.at<bfloat16>({b, h, m, 0}),
+                             d_scale,
+                             alibi_ptr + m * alibi_stride,
+                             attn_mask_ptr + m * attn_mask_stride,
+                             cmask_ptr + m * cmask_stride,
+                             select_nfltmax_at_0,
+                             ncausal,
+                             kv_len,
+                             attn_mask_prec,
+                             precision_of<T>::value);
+            }
+            bfloat16* w_ptr = &weight.at<bfloat16>({b, h, m_start, 0});
+            bfloat16* v_ptr = &present_value.at<bfloat16>({b, h / h_each_group_len, 0, 0});
+            float* fp32_out_ptr =
+                has_out_transpose ? &fp32_out.at<float>({b, m_start, h, 0}) : &fp32_out.at<float>({b, h, m_start, 0});
+            wv_gemm_ptr->executeGemm(m_cnt,
+                                     w_ptr,
+                                     v_ptr,
+                                     fp32_out_ptr,
+                                     wsp.data() + tid * wsp_size_per_thread,
+                                     wv_scratch_a ? &wv_scratch_a.at<bfloat16>({tid, 0}) : nullptr,
+                                     &wv_scratch_b.at<bfloat16>({b, h / h_each_group_len, 0}));
+        });
+        cpu_convert(&fp32_out.at<float>({0, 0, 0, 0}),
+                    &output_emb.at<bfloat16>({0, 0, 0, 0}),
+                    ov::element::f32,
+                    ov::element::bf16,
+                    B * H * q_len * head_size);
+    }
+
```

Comment:
Excellent work! 
A general question -
Can this multiquery executor work for FP32 too? If it is, we need use the template parameter instead of bfloat16 in the algorithm. If not, better to use a specific template implementation or exception check earlier to avoid running into problem when ENABLE_MLAS is switched off.

I know for FP32 case, MLAS is chosen. But I still suggest covering ONEDNN path too to provide another option (for cross-checking maybe) if possible.

Reply:
This brgemm executor here is designed for BF16xBF16->FP32 only. For FP32, MLAS has better performance as we test before. Even though combining them may clean up the code, more performance test is needed, and I am afraid we need extra PR to do so.

Reply:
another extra PR sounds okay. But we can have an exception check somewhere to avoid code running into this part and result into corruption.

Reply:
another question is, is there a situation where the output is bf16 so there is no need of cpu_convert at the end?

Reply:
oneDNN MHAkernel only works in BF16. It's fine to add a check, but is oneDNN kernel really used for FP32 ? Maybe oneDNN kernel should be specialized for BF16 only. @luo-cheng2021 What do you think ? Why do you leave oneDNN MHAkernel here for other precision?

https://github.com/openvinotoolkit/openvino/blob/6ad6ac7deef702b7bda0d46d5eaa196e20573823/src/plugins/intel_cpu/src/nodes/scaled_attn.cpp#L737